### PR TITLE
EPMEDU-761: Create "Enter Phone Number" page

### PR DESCRIPTION
**Pull request by Fileek**
- Implemented the screen for entering a phone number
- Added the app module for providing common dependencies
- Added StateViewModel to prevent duplicating StateFlow variables
- Rearranged parts of the login module
- Added data layer and di modules for both EnterCodeScreen and EnterPhoneScreen

| Light mode  | Dark mode  |
| ------------- | ------------- |
| <video src="https://user-images.githubusercontent.com/83027107/179358772-9b62d3df-fc68-4dd3-a0eb-59acc37377f3.mp4">  | <video src="https://user-images.githubusercontent.com/83027107/179358789-32c83f5d-f29f-4fab-a4d2-01cf1b8d95bf.mp4">|

### CODE REVIEW

File: feature/login/src/main/java/com/epmedu/animeal/login/code/data/EnterCodeRepositoryImpl.kt [hunk 1]
```diff
@@ -0,0 +1,19 @@
+package com.epmedu.animeal.login.code.data
+
+import android.content.SharedPreferences
+import com.epmedu.animeal.common.constants.SharedPreferencesKey.PHONE_NUMBER_KEY
+import javax.inject.Inject
+
+internal class EnterCodeRepositoryImpl @Inject constructor(
+    private val sharedPreferences: SharedPreferences
```
Comment (egorikftp) [inline feature/login/src/main/java/com/epmedu/animeal/login/code/data/EnterCodeRepositoryImpl.kt]: can we try to use `androidx datastore`?

File: feature/login/src/main/java/com/epmedu/animeal/login/code/di/EnterCodeModule.kt [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+import android.content.SharedPreferences
+import com.epmedu.animeal.login.code.data.EnterCodeRepository
+import com.epmedu.animeal.login.code.data.EnterCodeRepositoryImpl
+import dagger.Module
+import dagger.Provides
+import dagger.hilt.InstallIn
+import dagger.hilt.android.components.ViewModelComponent
+import dagger.hilt.android.scopes.ViewModelScoped
+
+@[Module InstallIn(ViewModelComponent::class)]
```
Comment (egorikftp) [inline feature/login/src/main/java/com/epmedu/animeal/login/code/di/EnterCodeModule.kt]: why not?

```kotlin
@Module
@InstallIn
```

File: feature/login/src/main/java/com/epmedu/animeal/login/code/domain/EnterCodeViewModel.kt [hunk 3]
```diff
@@ -1,19 +1,25 @@
...
-internal class EnterCodeViewModel : ViewModel() {
+@HiltViewModel
+internal class EnterCodeViewModel @Inject constructor(
+    repository: EnterCodeRepository
+) : ViewModel() {
 
     private val _model = MutableStateFlow(
         EnterCodeScreenModel(
-            phoneNumber = getPhoneNumber(),
+            phoneNumber = repository.getPhoneNumber(),
```
Comment (egorikftp) [inline feature/login/src/main/java/com/epmedu/animeal/login/code/domain/EnterCodeViewModel.kt]: better to make empty initial state and inside `init` block make some async operations to extract phone number

File: feature/login/src/main/java/com/epmedu/animeal/login/phone/domain/EnterPhoneViewModel.kt [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+    private val _model = MutableStateFlow(EnterPhoneModel())
+    val model: StateFlow<EnterPhoneModel> = _model.asStateFlow()
+
+    private val currentModel get() = model.value
+
+    fun updateNumber(newNumber: String) {
+        _model.value =
+            currentModel.copy(
+                phoneNumber = newNumber,
+                isNextEnabled = newNumber.length == PHONE_NUMBER_LENGTH
```
Comment (egorikftp) [inline feature/login/src/main/java/com/epmedu/animeal/login/phone/domain/EnterPhoneViewModel.kt]: can we also check by this one?

https://developer.android.com/reference/android/telephony/PhoneNumberUtils.html#isGlobalPhoneNumber(java.lang.String)

File: feature/login/src/main/java/com/epmedu/animeal/login/code/data/EnterCodeRepositoryImpl.kt (hunk 1 above)
Comment (Fileek) [reply feature/login/src/main/java/com/epmedu/animeal/login/code/data/EnterCodeRepositoryImpl.kt]: Changed

File: feature/login/src/main/java/com/epmedu/animeal/login/code/di/EnterCodeModule.kt (hunk 2 above)
Comment (Fileek) [reply feature/login/src/main/java/com/epmedu/animeal/login/code/di/EnterCodeModule.kt]: As you wish

File: feature/login/src/main/java/com/epmedu/animeal/login/code/domain/EnterCodeViewModel.kt (hunk 3 above)
Comment (Fileek) [reply feature/login/src/main/java/com/epmedu/animeal/login/code/domain/EnterCodeViewModel.kt]: Good point. Changed.

File: feature/login/src/main/java/com/epmedu/animeal/login/phone/domain/EnterPhoneViewModel.kt (hunk 4 above)
Comment (Fileek) [reply feature/login/src/main/java/com/epmedu/animeal/login/phone/domain/EnterPhoneViewModel.kt]: Done

File: library/common/src/main/java/com/epmedu/animeal/common/di/AppModule.kt [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+    @Singleton
+    @Provides
+    fun providesPreferencesDataStore(
+        @ApplicationContext context: Context
+    ): DataStore<Preferences> {
+        return PreferenceDataStoreFactory.create(
+            corruptionHandler = ReplaceFileCorruptionHandler(
+                produceNewData = { emptyPreferences() }
+            ),
+            migrations = listOf(SharedPreferencesMigration(context, DATA_STORE_PREFERENCES_KEY)),
```
Comment (egorikftp) [inline library/common/src/main/java/com/epmedu/animeal/common/di/AppModule.kt]: do we really need migrations? Previously we didn't have shared preferences

File: library/common/src/main/java/com/epmedu/animeal/common/di/AppModule.kt [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+            scope = CoroutineScope(Dispatchers.IO + SupervisorJob()),
```
Comment (egorikftp) [inline library/common/src/main/java/com/epmedu/animeal/common/di/AppModule.kt]: scope unnecessary, please check default parameter inside `PreferenceDataStoreFactory.create`

File: library/common/src/main/java/com/epmedu/animeal/common/di/AppModule.kt (hunk 5 above)
Comment (egorikftp) [reply library/common/src/main/java/com/epmedu/animeal/common/di/AppModule.kt]: the same for `produceFile`

Comment (Fileek) [reply library/common/src/main/java/com/epmedu/animeal/common/di/AppModule.kt]: Migrations removed. But `produceFile` is necessary

File: library/common/src/main/java/com/epmedu/animeal/common/di/AppModule.kt (hunk 6 above)
Comment (Fileek) [reply library/common/src/main/java/com/epmedu/animeal/common/di/AppModule.kt]: My bad. Removed.

File: feature/login/src/main/java/com/epmedu/animeal/login/code/domain/EnterCodeViewModel.kt [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+@HiltViewModel
+internal class EnterCodeViewModel @Inject constructor(
+    private val repository: EnterCodeRepository
+) : StateViewModel<EnterCodeState>(EnterCodeState()) {
+
+    private val _isCodeCorrect = MutableStateFlow(false)
+    val isCodeCorrect: StateFlow<Boolean> = _isCodeCorrect.asStateFlow()
+
+    init {
+        viewModelScope.launch {
```
Comment (egorikftp) [inline feature/login/src/main/java/com/epmedu/animeal/login/code/domain/EnterCodeViewModel.kt]: launch inside launch?)

maybe better `viewModelScope.launch{}` and `viewModelScope.launch {}`?

Comment (Fileek) [reply feature/login/src/main/java/com/epmedu/animeal/login/code/domain/EnterCodeViewModel.kt]: Ok)

**Review by egorikftp — APPROVED** [review]

**Review by Muhammadali-mk — APPROVED** [review]

**Pull request merged**